_posts/2019-06-05-uncertainty-estimates-tfprobability/uncertainty_tfprobability.Rmd

```diff
@@ -0,0 +1,782 @@
+```
+
+As with the simulated data, the linear models seems to "do the right thing". And here too, we think we will want to augment this with the spread in the data: Thus, on to way 3.
+
+### Single predictor: Combining both types
+
+Here we go. Again, `posterior_mean_field` and `prior_trainable` look just like in the epistemic-only case.
+
+```{r}
+model <- keras_model_sequential() %>%
+  layer_dense_variational(
+    units = 2,
+    make_posterior_fn = posterior_mean_field,
+    make_prior_fn = prior_trainable,
+    kl_weight = 1 / n,
+    activation = "linear"
+  ) %>%
+  layer_distribution_lambda(function(x)
+    tfd_normal(loc = x[, 1, drop = FALSE],
+               scale = 1e-3 + tf$math$softplus(0.01 * x[, 2, drop = FALSE])))
+
+
+negloglik <- function(y, model)
+  - (model %>% tfd_log_prob(y))
+model %>% compile(optimizer = optimizer_adam(lr = learning_rate), loss = negloglik)
+hist <-
+  model %>% fit(
+    X_train[, i, drop = FALSE],
+    y_train,
+    validation_data = list(X_val[, i, drop = FALSE], y_val),
+    epochs = n_epochs,
+    batch_size = batch_size
+  )
+
+yhats <- purrr::map(1:100, function(x)
+  model(tf$constant(X_val[, i, drop = FALSE])))
+means <-
+  purrr::map(yhats, purrr::compose(as.matrix, tfd_mean)) %>% abind::abind()
+sds <-
+  purrr::map(yhats, purrr::compose(as.matrix, tfd_stddev)) %>% abind::abind()
+
+means_gathered <- data.frame(cbind(X_val[, i], means)) %>%
+  gather(key = run, value = mean_val,-X1)
+sds_gathered <- data.frame(cbind(X_val[, i], sds)) %>%
+  gather(key = run, value = sd_val,-X1)
+
+lines <-
+  means_gathered %>% inner_join(sds_gathered, by = c("X1", "run"))
+
+mean <- apply(means, 1, mean)
+
+#lines <- lines %>% filter(run=="X3" | run =="X4")
+
+ggplot(data.frame(x = X_val[, i], y = y_val, mean = as.numeric(mean)), aes(x, y)) +
+  geom_point() +
+  theme(legend.position = "none") +
+  geom_line(aes(x = X_val[, i], y = mean), color = "violet", size = 1.5) +
+  geom_line(
+    data = lines,
+    aes(x = X1, y = mean_val, color = run),
+    alpha = 0.2,
+    size = 0.5
+  ) +
+geom_ribbon(
+  data = lines,
+  aes(
+    x = X1,
+    ymin = mean_val - 2 * sd_val,
+    ymax = mean_val + 2 * sd_val,
+    group = run
+  ),
+  alpha = 0.01,
+  fill = "grey",
+  inherit.aes = FALSE
+)
+
+```
+
+And the output?
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Combined uncertainty on the Combined Cycle Power Plant Data Set; single predictors."}
+knitr::include_graphics("images/uci_both.png")
+```
+
+This looks useful! Let's wrap up with our final test case: Using all four predictors together.
+
+### All predictors
+
+The training code used in this scenario looks just like before, apart from our feeding all predictors to the model. For plotting, we resort to displaying the first principal component on the x-axis -- this makes the plots look noisier than before. We also display fewer lines for the epistemic and epistemic plus aleatoric cases (20 instead of 100). Here are the results:
```

Comment:
```suggestion
The training code used in this scenario looks just like before, apart from our feeding all predictors to the model. For plotting, we resort to displaying the first principal component on the x-axis -- this makes the plots look noisier than before. We also display fewer lines for the epistemic and epistemic-plus-aleatoric cases (20 instead of 100). Here are the results:
```

Reply:
^^ added hyphens since it's being used as a compound modifier for "cases"

Reply:
oh, I was wondering how to write that, that's nice!